projectns/ns_listgroupcloaks.c

```diff
@@ -0,0 +1,134 @@
+ */
+
+#include "fn-compat.h"
+#include "atheme.h"
+#include "projectns.h"
+
+static void cmd_listgroupcloaks(sourceinfo_t *si, int parc, char *parv[]);
+
+command_t ns_listgroupcloaks = { "LISTGROUPCLOAKS", N_("List accounts with cloaks belonging to your projects."), AC_AUTHENTICATED, 2, cmd_listgroupcloaks, { .path = "freenode/ns_listgroupcloaks" } };
+
+struct cloak_ns {
+	struct projectns *project;
+	char pattern[BUFSIZE];
+};
+
+static void cmd_listgroupcloaks(sourceinfo_t *si, int parc, char *parv[])
+{
+	char *filter = parv[0];
+
+	mowgli_node_t *n, *tn;
+	mowgli_list_t *plist = projectsvs->myuser_get_projects(si->smu);
+	unsigned int num_projects = 0;
+	mowgli_list_t cloak_ns_list = { NULL, NULL, 0 };
+
+	MOWGLI_ITER_FOREACH(n, plist->head)
+	{
+		struct project_contact *contact = n->data;
+		mowgli_node_t *n2;
+
+		MOWGLI_ITER_FOREACH(n2, contact->project->cloak_ns.head)
+		{
+			struct cloak_ns *cns = smalloc(sizeof *cns);
+			cns->project = contact->project;
+			mowgli_strlcpy(cns->pattern, (const char *)n2->data, sizeof cns->pattern);
+			mowgli_strlcat(cns->pattern, "/*", sizeof cns->pattern);
+			mowgli_node_add(cns, mowgli_node_create(), &cloak_ns_list);
+		}
+
+		num_projects++;
+	}
+
+	if (!num_projects)
+	{
+		command_fail(si, fault_noprivs, _("You are not an authorized group contact for any project."));
+		return;
+	}
+
+	if (filter)
+		command_success_nodata(si, _("Assigned cloaks in your projects matching \2%s\2:"), filter);
+	else
+		command_success_nodata(si, _("Assigned cloaks in your projects:"));
+
+	unsigned int matches = 0;
+	struct myentity *mt;
+	struct myuser *mu;
+	struct metadata *md;
+	struct myentity_iteration_state state;
+
+	MYENTITY_FOREACH_T(mt, &state, ENT_USER)
+	{
+		mu = user(mt);
+		md = metadata_find(mu, "private:usercloak");
+		if (md == NULL)
+			continue;
+
+		if (filter != NULL && match(filter, md->value))
+			continue;
+
+		MOWGLI_ITER_FOREACH(n, cloak_ns_list.head)
+		{
+			struct cloak_ns *cns = (struct cloak_ns*)n->data;
+			if (!match(cns->pattern, md->value))
+			{
+				command_success_nodata(si, "- %s (%s) [%s]", md->value, mt->name, cns->project->name);
+				matches++;
+				break;
+			}
+		}
+	}
+
+	if (matches == 0)
+	{
+		if (filter)
+			command_success_nodata(si, _("No assigned cloaks matched pattern \2%s\2"), filter);
+		else
+			command_success_nodata(si, _("There are no assigned cloaks in your projects."));
+	}
+	else
+	{
+		if (filter)
+			command_success_nodata(si,
+				ngettext(N_("\2%d\2 match for pattern \2%s\2"), N_("\2%d\2 matches for pattern \2%s\2"), matches),
+				matches, filter);
+		else
+			command_success_nodata(si, ngettext(N_("\2%d\2 match"), N_("\2%d\2 matches"), matches), matches);
+	}
+
+	if (filter)
+		logcommand(si, CMDLOG_GET, "LISTGROUPCLOAKS: \2%s\2", filter);
+	else
+		logcommand(si, CMDLOG_GET, "LISTGROUPCLOAKS");
+
+	MOWGLI_ITER_FOREACH_SAFE(n, tn, cloak_ns_list.head)
+	{
+		mowgli_node_delete(n, &cloak_ns_list);
```

Comment:
There's no need to delete nodes from a list that only exists on the stack. It will be invalidated when the function returns anyway, so whether it has head, tail and count members set is irrelevant. You still need the `sfree` and `mowgli_node_free` though.